Mathlib/Tactic/Algebraize.lean

```diff
@@ -0,0 +1,264 @@
+and then returning `Algebra.Property`. -/
+def algebraizeGetParam (thm : Name) (stx : Syntax) : AttrM Name := do
+  match stx with
+  | `(attr| algebraize $name:ident) => return name.getId
+  /- If no argument is provided, assume `thm` is of the form `RingHom.Property`,
+  and return `Algebra.Property` -/
+  | `(attr| algebraize) =>
+    match thm with
+    | .str `RingHom t => return .str `Algebra t
+    | _ =>
+      throwError "theorem name must be of the form `RingHom.Property` if no argument is provided"
+  | _ => throwError "unexpected algebraize argument"
+
+/-- A user attribute that is used to tag `RingHom` properties that can be converted to `Algebra`
+properties. Using an (optional) parameter, it will also generate a `Name` of a declaration which
+will help the `algebraize` tactic access the corresponding `Algebra` property.
+
+There are two cases for what declaration corresponding to this `Name` can be.
+1. An inductive type (i.e. the `Algebra` property itself), in this case it is assumed that the
+`RingHom` and the `Algebra` property are definitionally the same, and the tactic will construct the
+`Algebra` property by giving the `RingHom` property as a term.
+
+2. A constructor for the `Algebra` property. In this case it is assumed that the `RingHom` property
+is the last argument of the constructor, and that no other explicit argument is needed. The tactic
+then constructs the `Algebra` property by applying the constructor to the `RingHom` property.
+
+Finally, if no argument is provided to the `algebraize` attribute, it is assumed that the tagged
+declaration has name `RingHom.Property` and that the corresponding `Algebra` property has name
+`Algebra.Property`. The attribute then returns `Algebra.Property` (so assume case 1 above). -/
+initialize algebraizeAttr : ParametricAttribute Name ←
+  registerParametricAttribute {
+    name := `algebraize,
+    descr :=
+"Tag that lets the `algebraize` tactic know which `Algebra` property corresponds to this `RingHom`
+property.",
+    getParam := algebraizeGetParam }
+
+end Lean.Attr
+
+namespace Mathlib.Tactic
+
+namespace Algebraize
+
+/-- Given an expression `f` of type `RingHom A B` where `A` and `B` are commutative semirings,
+this function adds the instance `Algebra A B` to the context (if it does not already exist).
+
+This function also requries the type of `f`, given by the parameter `ft`. The reason this is done
+(even though `ft` can be inferred from `f`) is to avoid recomputing `ft` in the `algebraize` tactic,
+as when `algebraize` calls `addAlgebraInstanceFromRingHom` it has already computed `ft`. -/
+def addAlgebraInstanceFromRingHom (f ft : Expr) : TacticM Unit := withMainContext do
+  let (_, l) := ft.getAppFnArgs
+  -- The type of the corresponding algebra instance
+  let alg ← mkAppOptM `Algebra #[l[0]!, l[1]!, none, none]
+  try let _ ← synthInstance alg -- If the instance already exists, we do not do anything
```

Comment:
@joneugster I just had the following thought: Is this too expensive? Should I maybe instead loop through the local context and check that nothing already has this type?

Reply:
I meant that the typical case is that this will fail to synthesize, which might still be expensive? Although I guess that it could be synthesized whilst not being a local hypothesis, so I guess the above check has to be done.